mava/networks/utils/sable/discrete_trainer_executor.py

```diff
@@ -0,0 +1,337 @@
+        )
+        logit = logit.at[:, start_idx:end_idx].set(chunk_logit)
+
+    return logit, hstates
+
+
+def get_shifted_actions(action: chex.Array, legal_actions: chex.Array, n_agents: int) -> chex.Array:
+    """Get the shifted action sequence for predicting the next action."""
+    # Get the batch size, sequence length, and action dimension
+    batch_size, sequence_size, action_dim = legal_actions.shape
+
+    # Create a shifted action sequence for predicting the next action
+    # Initialize the shifted action sequence.
+    shifted_actions = jnp.zeros((batch_size, sequence_size, action_dim + 1))
+
+    # Set the start-of-timestep token (first action as a "start" signal)
+    start_timestep_token = jnp.zeros(action_dim + 1).at[0].set(1)
+
+    # One hot encode the action
+    one_hot_action = jax.nn.one_hot(action, action_dim)
+
+    # Insert one-hot encoded actions into shifted array, shifting by 1 position
+    shifted_actions = shifted_actions.at[:, :, 1:].set(one_hot_action)
+    shifted_actions = jnp.roll(shifted_actions, shift=1, axis=1)
+
+    # Set the start token for the first agent in each timestep
+    shifted_actions = shifted_actions.at[:, ::n_agents, :].set(start_timestep_token)
+
+    return shifted_actions
+
+
+def init_sable(
+    encoder: nn.Module,
+    decoder: nn.Module,
+    obs_carry: chex.Array,
+    hstates: chex.Array,
+    key: chex.PRNGKey,
+) -> chex.Array:
+    """Initializating the network: Applying non chunkwise encoding-decoding."""
+    # Get the observation, legal actions, and timestep id
+    obs, legal_actions, timestep_id = (
+        obs_carry.agents_view,
+        obs_carry.action_mask,
+        obs_carry.step_count,
+    )
+
+    # Apply the encoder
+    v_loc, obs_rep, _ = execute_encoder_parallel(
+        encoder=encoder, obs=obs, decayed_hstate=hstates[0], timestep_id=timestep_id
+    )
+
+    # Apply the decoder
+    _ = autoregressive_act(
+        decoder=decoder,
+        obs_rep=obs_rep,
+        legal_actions=legal_actions,
+        hstates=hstates[1],
+        timestep_id=timestep_id,
+        key=key,
+    )
+
+    return v_loc
+
+
+def execute_encoder_parallel(
+    encoder: nn.Module,
+    obs: chex.Array,
+    decayed_hstate: chex.Array,
+    timestep_id: chex.Array,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Parallel encoding for discrete action spaces."""
+    # Apply the encoder
+    v_loc, obs_rep, updated_hstate = encoder.recurrent(obs, decayed_hstate, timestep_id)
+    return v_loc, obs_rep, updated_hstate
+
+
+def execute_encoder_chunkwise(
+    encoder: nn.Module,
+    obs: chex.Array,
+    decayed_hstate: chex.Array,
+    timestep_id: chex.Array,
+    chunk_size: int,
+) -> Tuple[chex.Array, chex.Array, chex.Array]:
+    """Chunkwise encoding for Non-memory Sable and for discrete action spaces."""
+    # Get the batch and sequence dimensions
+    batch_dim, agents_per_seq = obs.shape[:2]
+    # Initialize the value location and observation representation
```

Comment:
I think these comments are unecessary, it is clear what is happening. Also can you please use `B, N = ...` for the shapes
```suggestion
    batch_dim, agents_per_seq = obs.shape[:2]
```